Introduce testing system with pgTAP

fljdin opened the pull request:
Sequel to https://github.com/cybertec-postgresql/db_migrator/pull/27

If the testing code base seems good to you, please squash commits into one.

A next work will be moving common regression tests from `ora_migrator`  to this system and leaving all oracle plugin-related tests untouched.

fljdin committed:
Introduce testing system with pgTAP

Add a no-operation (noop) plugin that rely on file_fdw and /dev/null. Setup script
executes db_migrate_prepare function and populates staging tables with convenient
datasets needed by pgTAP suite.

fljdin committed:
Test tables migration

Provide a "tables" testing file with simple assertions over
"db_migrate_mkforeign" and "db_migrate_tables" methods.

Link foreign tables to test/data directory by overriding noop_mkforeign method
on plugin install.

fljdin committed:
Test indexes and constraints migration

Provide "indexes" and "constraints" testing file. Change Makefile to fix tests
ordering. All testing files are now committed to perform a complete migration.

fljdin committed:
Add CONTRIBUTING.md

Provide a documentation about different ways to contribute and submit pull
requests. Explain how to run testing suite locally.

Use of schedule files with pg_regress to run setup scripts before any tests.

fljdin committed:
Test partitions migration

Provide a "partitions" testing file with a new schema "part1" and new table
"pt1" and its datafile.

fljdin committed:
Applied suggestions by laurenz

Install noop_migrator.control
Remove PGXN reference
Newline at the end files

laurenz committed:
Improve contributor documentation

Remove the paragraph that specifies which tags a new contributor
should work on.  We are not yet operating at the scale where this
would be necessary.  Let's add rules when we need them.

Other than that, I tried to improve the grammar, mostly by adding
articles.

This addresses #27#discussion_r1280738187

laurenz committed:
Improve contributor documentation

Remove the paragraph that specifies which tags a new contributor
should work on.  We are not yet operating at the scale where this
would be necessary.  Let's add rules when we need them.

Other than that, I tried to improve the grammar, mostly by adding
articles.

This addresses #27#discussion_r1280738187

fljdin committed:
Applied suggestions by laurenz

Set client_min_messages to warning

laurenz committed:
Merge remote-tracking branch 'fljdin/testing-with-pgtap' into testing-with-pgtap

laurenz committed:
Merge remote-tracking branch 'refs/remotes/fljdin/testing-with-pgtap' into testing-with-pgtap

laurenz committed:
Merge branch 'master' into testing-with-pgtap

laurenz committed:
Small improvements for the regression tests

- Add a note to the change log

- remove unnecessary psql settings

- fix whitespace

laurenz reviewed (changes requested):
I notice that some features are untested:

- subpartitions
- views
- sequences
- functions
- triggers

I think it would be great if you added some tests for these as well.

fljdin committed:
Test subpartitions migration

Use RANGE and HASH type to migrate a new testing partitioned table part1.pt2

fljdin committed:
Test views, functions and triggers migration

Use a classic INSTEAD OF trigger on a view with two functions (multiply and
divide)

fljdin committed:
Test schemas and sequences migration

As schemas and sequences are created by db_migrate_mkforeign(), new tests are
provided inside tables.sql testing file

laurenz committed:
Code review for new pgTAP tests

Move the trigger definition to the COPY input.
Fix some white space.

laurenz commented:
Thanks for all the effort!
